基本機能実装 #3

isoden opened the pull request.

isoden committed:
基本機能実装 #3

isoden committed:
天地中央実装 #3

haribote commented on demo/index.slim:
```diff
@@ -3,7 +3,17 @@ html
   body
-    #contents
+    a.maboroshiBox href="#contents" open!
+    #contents.is-hide
```
**#contents** というIDはサンプルとして微妙なので変えましょう。

haribote commented on demo/index.slim:
```diff
@@ -3,7 +3,17 @@ html
\ No newline at end of file
+      p hogehoge
+    script src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"
+    script src="../dist/jquery.maboroshiBox.js"
```
CSSとJSでパスの指定方法が違うのは気持ち悪いです。

haribote commented on demo/site-assets/scss/maboroshiBox.scss:
```diff
@@ -0,0 +1,22 @@
+.maboroshiBox-modal {
```
BEMでお願いします。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,123 @@
+    this.el        = el;
+    this.$el       = $(this.el);
+    this.options   = options;
+    this.$target   = $(this.$el.attr('href')).addClass('maboroshiBox-modal');
```
ターゲットを拾ってくるのは↓で十分です。

```
$(this.el.hash)
```

addClassをここでやりたい気持ちは分かるけど、initメソッドでやりましょう。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,123 @@
+    */
+    init: function () {
+      // 背景レイヤー設定
+      this.$bgLayer = $('<div class="maboroshiBox-bgLayer"/>').appendTo('body');
```
モーダルウィンドウの数だけ背景レイヤーが生成されてしまうので効率的でない。
何をどの単位でオブジェクト&インスタンス化すべきか、よく考えてください。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,123 @@
+     * @method createCloseButton
+     */
+    createCloseButton: function () {
+      this.closeButton = $(this.options.closeHtml);
```
プロパティの命名規則が他とオブジェクトと違います。
さらに、コンストラクターが呼ばれるたびにjQueryオブジェクトを生成しているので、これもプラグインとしてどのように管理すべきかを考えましょう。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,123 @@
+     * イベントを紐付けるためのメソッド
+     * @method addEvent
+     */
+    addEvent: function () {
```
addEventってメソッド名が微妙な気がする。
他のプラグインだとどうなっているかを参考にしてください。

isoden committed:
issue #9 の対応 #3

\#contents -> #modal1
css, js のパス変更
BEMに変更
ターゲットを$(this.el.hash)に変更
initメソッド, createCloseButtonメソッドを初回のみ実行するように変更
this.closeButton -> this.$closeButtonに変更
addEvent -> bindEventに変更

isoden commented:
動作不備あったので一旦プルリク下げます

isoden committed:
スペルミス修正、 initメソッド変更 #3

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,133 @@
   // コンストラクタ定義
-  var MaboroshiBox = function () {
+  var MaboroshiBox = function (el, options) {
+    // プロパティ定義
```
MaboroshiBoxオブジェクトがどういう単位のオブジェクトなのかを設計することが大事。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,133 @@
     options = $.extend({}, $.fn.maboroshiBox.defaults, options);
 
     return this.each(function (i, el) {
+      var maboroshiBox = new MaboroshiBox(el, options);
```
MaboroshiBoxオブジェクトがjQueryプラグインを引き当てた要素（とターゲットとなる要素）を1単位とするのであれば、MaboroshiBoxオブジェクトのインスタンスは複数生成されるので、それらのコレクション（及び共有する要素）を取りまとめるコントローラーが必要。
とりまとめ役となるコントローラーはMaboroshiBoxオブジェクト内に実装されるべきではなくて、別のオブジェクトとして定義した方が良い。
で、MaboroshiBoxオブジェクトのインスタンスを生成するときにコントローラーオブジェクトのインスタンスを渡してやれば、MaboroshiBoxオブジェクト内からコントローラオブジェクトのメソッドを呼んだりできる。
コンストラクターは1個でなくてはいけないわけじゃないので、オブジェクトの役割や単位に応じて定義すれば良い。

haribote replied:
> で、MaboroshiBoxオブジェクトのインスタンスを生成するときにコントローラーオブジェクトのインスタンスを渡してやれば、MaboroshiBoxオブジェクト内からコントローラオブジェクトのメソッドを呼んだりできる。

逆ですね、コントローラーオブジェクトにMaboroshiBoxインスタンスをプッシュしていってコレクションを管理できるようにした方が良い。

haribote commented on dist/jquery.maboroshiBox.js:
```diff
@@ -6,24 +6,133 @@
 
   // メソッド定義
   MaboroshiBox.prototype = {
+
```
prototypeプロパティを新しいオブジェクトで上書きしてしまうのはbadなんだってさ！

> Assign methods to the prototype object, instead of overwriting the prototype with a new object.

https://github.com/airbnb/javascript#constructors

isoden committed:
メソッド修正 #3

$bgLayerが何度も生成されてたので修正、それにあわせて他のメソッドとプロパティも修正
